add usage examples

WilliamHYZhang opened the pull request:
added usage examples for tf.image.rgb_to_grayscale and tf.image.grayscale_to_rgb

WilliamHYZhang committed:
add usage examples

added usage examples for tf.image.rgb_to_grayscale and tf.image.grayscale_to_rgb

alextp commented on tensorflow/python/ops/image_ops_impl.py:
````diff
@@ -1874,6 +1874,13 @@ def rgb_to_grayscale(images, name=None):
+    
+  Usage Example:
+   ```python
+   import tensorflow as tf
````
No need for importing tensorflow

WilliamHYZhang replied:
Ok, I will get rid of that. Thanks for the feedback.

mihaimaruseac reviewed (changes requested):
Please indent the code blocks in the documentation the same way in both blocks

alextp commented on tensorflow/python/ops/image_ops_impl.py:
````diff
@@ -1874,6 +1874,13 @@ def rgb_to_grayscale(images, name=None):
+   ```python
+   import tensorflow as tf
+   original = tf.random.normal(shape=(256, 256, 3))
+   converted = tf.image.rgb_to_grayscale(original)
````
Can you choose a sufficiently small image with nice floating point values (a 1x1 image would work here, for example) so we can make this a doctest?

WilliamHYZhang replied:
Ok, shall I replace line 1881 with `original = tf.random.normal(shape=(1, 1, 3))` instead? Default output value should already be tf.dtypes.float32, would that need to change as well? Thanks for the feedback.

WilliamHYZhang committed:
update usage examples

WilliamHYZhang commented:
@alextp @mihaimaruseac I made the requested changes. Thanks!

WilliamHYZhang committed:
fix indentation

mihaimaruseac commented:
I think @alextp also wanted you to print the value of `converted`, basically to make this a [Python doctest](https://docs.python.org/2/library/doctest.html). See for example https://github.com/tensorflow/tensorflow/blob/f70c46f8cd2a91b390455827cda65b5b5fe92ef0/tensorflow/python/framework/ops.py#L446-L468

alextp reviewed (changes requested):
Now can you add >>> to make these doctests? And replace the random input with a constant array to make the test deterministic?

WilliamHYZhang committed:
update doctests

WilliamHYZhang commented:
Would you like me to also add `print(converted)` as @mihaimaruseac suggested? When doing that, however, I get an output such as `Tensor("grayscale_to_rgb_53:0", shape=(1, 3, 3), dtype=float32)` but the two digits after the underscore ("53" in this case) changes each run.

mihaimaruseac commented:
We can print only the shape. See the doctest example in https://github.com/tensorflow/tensorflow/blob/f70c46f8cd2a91b390455827cda65b5b5fe92ef0/tensorflow/python/framework/ops.py#L446-L468

WilliamHYZhang committed:
fix doctests

WilliamHYZhang commented:
@mihaimaruseac thanks for the information, pushed the fix.

mihaimaruseac commented:
One more thing: `>>>` should be followed by a space

WilliamHYZhang committed:
add space after >>> for doctests

WilliamHYZhang commented:
Ok, I updated that in the PR. Thanks.

mihaimaruseac reviewed (commented):
Last thing (apologies for not catching this in the previous review): Please remove `Usage Example` (as it is redundant) and remove one indentation level from the code blocks in the doctest.

WilliamHYZhang committed:
remove "usage example" text, fix indentation

WilliamHYZhang commented:
No problem, pushed with the fix again. :)

mihaimaruseac reviewed (dismissed):
Thank you. We'll take it from here

gbaned commented:
@WilliamHYZhang Could you please address Ubuntu Sanity errors? Thanks!

WilliamHYZhang commented:
Hm, that's odd. When I run the example in Google Colab (https://colab.research.google.com/drive/1S2g8mhOj-RZDmc3JvJW7FQqM-_ds4a6V), no error returns. What can I do to address these errors in Ubuntu Sanity?

mihaimaruseac commented:
I don't understand why the log is not visible, let's try running the test again.

In the meanwhile, can you check with pylint if the code doesn't result in formatting errors?

mihaimaruseac commented:
Probably you have to move the doctest before the `Args:` section, not after `Returns:`

WilliamHYZhang committed:
move doctests before args: section

WilliamHYZhang commented:
Ok, pushed again. I moved the doctests like you suggested, thanks!

WilliamHYZhang committed:
doctests fix ubuntu cpu test fail

mihaimaruseac reviewed (commented):
Instead of printing the shape (which is not giving enough information), can you print `converted.numpy()`? Here's my interactive session:

```console
>>> import tensorflow as tf
>>> original = tf.constant([[[1.0, 2.0, 3.0]]])    
>>> converted = tf.image.rgb_to_grayscale(original)
>>> converted
<tf.Tensor: shape=(1, 1, 1), dtype=float32, numpy=array([[[1.8149]]], dtype=float32)>
>>> converted.numpy()
array([[[1.8149]]], dtype=float32)
>>> converted.shape
TensorShape([1, 1, 1])
```

WilliamHYZhang committed:
doctests update

print converted.numpy() instead of shape for rgb_to_grayscale and grayscale_to_rgb

WilliamHYZhang commented:
@mihaimaruseac thank you for the suggestion, I pushed the update.

WilliamHYZhang committed:
doctest fix ubuntu cpu test

fix expected return for doctests with grayscale_to_rgb, rgb_to_grayscale

WilliamHYZhang committed:
add newline after doctest expected return

WilliamHYZhang commented:
Oh man, seems like the Ubuntu CPU test is failing again. Checked the status and seems like it returns ```[[[1.8149]]]``` instead of the expected ```array([[[1.8149]]], dtype=float32)```. Updated fix to reflect that. Also, it looks like it also was expecting the final ``` in the return as well, so I added a newline to hopefully fix that as well.

alextp commented on tensorflow/python/ops/image_ops_impl.py:
```diff
@@ -1866,7 +1866,15 @@ def rgb_to_grayscale(images, name=None):
+  >>> original = tf.constant([[[1.0, 2.0, 3.0]]])
+  >>> converted = tf.image.rgb_to_grayscale(original)
+  >>> print(converted.numpy())
+  [[[1.8149]]]
```
Maybe to 1.81... to avoid using too many digits of precision?

WilliamHYZhang committed:
doctest remove commas from array return

fix yet another Ubuntu CPU test fail

WilliamHYZhang commented:
CPU test failed again :(
Seems like it didn't return the commas in the np array for grayscale_to_rgb, just pushed to fix that. @alextp Is 4 decimal places of precision too much? Should I change that line to [[[1.81...]]], I don't know if that's syntactically correct. Thanks

mihaimaruseac commented:
Yes, please convert to `[[[1.81...]` as `...` in doctest mean "don't try to match after this". 

Sorry it's taking so long and so many review rounds

WilliamHYZhang committed:
doctest update decimal precision

WilliamHYZhang commented:
All tests now passing besides ```import/copybara — An error happened while migrating the change```, what does this mean?

mihaimaruseac commented:
Let's try it again. I think it's an internal error.

WilliamHYZhang commented:
Strange, the import/copybara error seems to be still happening.

WilliamHYZhang commented:
@mihaimaruseac Shall I maybe refork the repo and create a new PR for this?

mihaimaruseac commented:
No need to. There is an error on our side, there are people looking into it.

Sorry it's taking so long and sorry I couldn't respond earlier
